### PR TITLE
More zalloc: fix default initialisation of some allocated objects

### PR DESCRIPTION
**Pull request by FdaSilvaYY**
1st commit is intended to be back-ported.
So, I revert some of his changes in the second commit directly.

### CODE REVIEW

File: crypto/asn1/bio_ndef.c [hunk 1]
```diff
@@ -135,6 +135,7 @@ BIO *BIO_new_NDEF(BIO *out, ASN1_VALUE *val, const ASN1_ITEM *it)
     ndef_aux->ndef_bio = sarg.ndef_bio;
     ndef_aux->boundary = sarg.boundary;
     ndef_aux->out = out;
+    ndef_aux->derbuf = NULL;
```
Comment (richsalz) [inline crypto/asn1/bio_ndef.c]: not needed because of the zalloc call.

File: crypto/asn1/bio_asn1.c [hunk 2]
```diff
@@ -145,7 +145,8 @@ const BIO_METHOD *BIO_f_asn1(void)
 static int asn1_bio_new(BIO *b)
 {
     BIO_ASN1_BUF_CTX *ctx;
-    ctx = OPENSSL_malloc(sizeof(*ctx));
+
+    ctx = OPENSSL_zalloc(sizeof(*ctx));
```
Comment (richsalz) [inline crypto/asn1/bio_asn1.c]: merge this up with the declaration line.

File: crypto/asn1/bio_asn1.c [hunk 3]
```diff
@@ -164,15 +165,10 @@ static int asn1_bio_init(BIO_ASN1_BUF_CTX *ctx, int size)
...
-    ctx->bufpos = 0;
-    ctx->buflen = 0;
-    ctx->copylen = 0;
     ctx->asn1_class = V_ASN1_UNIVERSAL;
     ctx->asn1_tag = V_ASN1_OCTET_STRING;
-    ctx->ex_buf = 0;
-    ctx->ex_pos = 0;
-    ctx->ex_len = 0;
     ctx->state = ASN1_STATE_START;
+
```
Comment (richsalz) [inline crypto/asn1/bio_asn1.c]: nope, don't add this line.

File: crypto/asn1/a_strnid.c [hunk 4]
```diff
@@ -256,6 +256,7 @@ static ASN1_STRING_TABLE *stable_get(int nid)
         rv->mask = tmp->mask;
         rv->flags = tmp->flags | STABLE_FLAGS_MALLOC;
     } else {
+        rv->nid = NID_undef;
```
Comment (richsalz) [inline crypto/asn1/a_strnid.c]: if you replace the malloc above with zalloc, then this line (and setting rv->make) aren't needed.

File: crypto/asn1/tasn_scn.c [hunk 5]
```diff
@@ -74,7 +74,8 @@
 ASN1_SCTX *ASN1_SCTX_new(int (*scan_cb) (ASN1_SCTX *ctx))
 {
     ASN1_SCTX *ret;
-    ret = OPENSSL_malloc(sizeof(*ret));
+
+    ret = OPENSSL_zalloc(sizeof(*ret));
```
Comment (richsalz) [inline crypto/asn1/tasn_scn.c]: merge this up into the declaration.

File: crypto/bio/bss_bio.c [hunk 6]
```diff
@@ -121,14 +121,12 @@ static int bio_new(BIO *bio)
 {
     struct bio_bio_st *b;
 
-    b = OPENSSL_malloc(sizeof(*b));
+    b = OPENSSL_zalloc(sizeof(*b));
```
Comment (richsalz) [inline crypto/bio/bss_bio.c]: merge this up into the declaration.

**Comment by richsalz** [comment]
To make sure I get it:  the first commit is for 1.0.2/1.0.1 and the first and second are for master?
+1

**Comment by FdaSilvaYY** [comment]
Yes, that's it.
The second commit is only for the master.

File: crypto/bio/bss_bio.c [hunk 7]
```diff
@@ -119,16 +119,13 @@ struct bio_bio_st {
...
 static int bio_new(BIO *bio)
 {
-    struct bio_bio_st *b;
+    struct bio_bio_st *b = OPENSSL_zalloc(sizeof(*b));
 
-    b = OPENSSL_malloc(sizeof(*b));
     if (b == NULL)
         return 0;
 
-    b->peer = NULL;
```
Comment (levitte) [inline crypto/bio/bss_bio.c]: No.  The null pointer isn't necessarily all 0 bits.  See http://c-faq.com/null/runtime0.html

File: crypto/bio/bss_bio.c [hunk 8]
```diff
@@ -119,16 +119,13 @@ struct bio_bio_st {
...
     /* enough for one TLS record (just a default) */
     b->size = 17 * 1024;
-    b->buf = NULL;
```
Comment (levitte) [inline crypto/bio/bss_bio.c]: Likewise, keep this line.

File: crypto/bio/bss_bio.c (hunk 7 above)
Comment (levitte) [inline crypto/bio/bss_bio.c]: That being said, I may have missed this in early similar PRs...

Comment (richsalz) [inline crypto/bio/bss_bio.c]: No, we only run on platforms where NULL is all-bits-zero.  See test/nptest.c

File: crypto/bio/bss_bio.c (hunk 8 above)
Comment (richsalz) [inline crypto/bio/bss_bio.c]: No.

File: crypto/bio/bss_bio.c (hunk 7 above)
Comment (richsalz) [inline crypto/bio/bss_bio.c]: You missed test/nptest :)

Comment (levitte) [inline crypto/bio/bss_bio.c]: Ah!

Ok then

**Comment by mattcaswell** [comment]
Ping @levitte

**Comment by mattcaswell** [comment]
@levitte is this good to go?

**Comment by levitte** [comment]
Yes, looks good to me

**Comment by mattcaswell** [comment]
Sorry, @levitte - was that a plus one?

**Comment by levitte** [comment]
Yes, sorry for the confusion.  +1

**Comment by mattcaswell** [comment]
Adding review label for @levitte. @richsalz will you merge?

**Comment by richsalz** [comment]
done.

**Pull request closed without merging**